sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h

```diff
@@ -0,0 +1,160 @@
+namespace sdk
+{
+namespace metrics
+{
+template<class T>
+class HistogramAggregator final : public Aggregator<T>
+{
+    
+public:
+    
+    /**
+     * Constructor for the histogram aggregator.  A sorted vector of boundaries is expected and boundaries are
+     * doubles regardless of the aggregator's templated data type.
+     *
+     * Sum is stored in values_[0]
+     * Count is stored in position_[1]
+     */
+    HistogramAggregator(metrics_api::BoundInstrumentKind kind, std::vector<double> boundaries)
+    {
+        this->kind_ = kind;
+        boundaries_   = boundaries;
+        this->values_     = std::vector<T>(2, 0);
+        this->checkpoint_ = std::vector<T>(2, 0);
+        bucketCounts_ = std::vector<int>(boundaries_.size() + 1, 0);
+    }
+    
+    /**
+     * Recieves a captured value from the instrument and inserts it into the current histogram counts.
+     * Depending on the use case, linear search or binary search based implementations may be preferred.
+     * In a uniformly distributed dataset, linear search outperforms binary search until 512 buckets.
+     *
+     * @param val, the raw value used in aggregation
+     * @return none
+     */
+    void update(T val) override
+    {
+        
+        int bucketID = boundaries_.size();
+        for (int i = 0; i < boundaries_.size(); i++)
+        {
+            if (val < boundaries_[i]) // concurrent read is thread-safe
+            {
+                bucketID = i;
+                break;
+            }
+        }
+        
+        // Alternate implementation with binary search
+        //auto pos = std::lower_bound (boundaries_.begin(), boundaries_.end(), val);
+        //bucketCounts_[pos-boundaries_.begin()] += 1;
+        
+        this->mu_.lock();
+        this->values_[0] += val;
+        this->values_[1] += 1;
+        bucketCounts_[bucketID] += 1;
+        this->mu_.unlock();
+    }
+    
+    /**
+     * Checkpoints the current value.  This function will overwrite the current checkpoint with the
+     * current value.
+     *
+     * @param none
+     * @return none
+     */
+    void checkpoint() override
+    {
+        this->checkpoint_ = this->values_;
+        this->values_[0]=0;
+        this->values_[1]=0;
+    }
+    
+    /**
+     * Merges the values of two aggregators in a semantically accurate manner.
+     * A histogram aggregator can only be merged with another histogram aggregatos with the same boudnaries.
+     * A histogram merge first adds the sum and count values then iterates over the adds the bucket counts
+     * element by element.
+     *
+     * @param other, the aggregator with merge with
+     * @return none
+     */
+    void merge(HistogramAggregator other)
+    {
+        this->mu_.lock();
+        this->values_[0] += other.values_[0];
+        this->values_[1] += other.values_[1];
+        
+        for (int i = 0; i < bucketCounts_.size(); i++)
+        {
+            bucketCounts_[i] += other.bucketCounts_[i];
```

Comment:
Who should be responsible of making sure the `other.bucketCounts` won't overflow? Do we in general want to check the buckets are the same before proceeding?

Reply:
This is a good suggestion.  When you say " `other.bucketCounts` won't overflow," are you referring to a scenario in which adding the values together would exceed the capacity of the data type?

Reply:
For example, if `other.bucketCounts_.size() <= i`.

Reply:
There are two separate things:
1. index overflow while accessing the bucketCounts_.
2. value overflow while performing `+=`.

1 is more scary.

Reply:
I preempted the possibility of an index out of bounds issue by ensuring that the boundary vectors of both aggregators are identical.  

For the value overflow, I can think of some hacky ways to do it, for example if abs(a)+abs(b) < 0 to indicate a wraparound.  I could also subtract from a variable set to int_max.  I feel like there has to be a native c++ tool/function which does this but I can't seem to find it...